### PR TITLE
[manager] support selecting multiple V6D peers

### PR DESCRIPTION
**Pull request by lucky-zzz**
背景

GetCacheLocationsByBackend 目前每个 V6D selector 最多只能选择一个 peer。当缓存块分散在多个 peer 上时，这会降低可命中范围。

实现

- 为 BackendLocationSelector 增加 max_peer_count，要求该值大于 0。
- CacheManager 校验参数，并限制该字段仅用于 V6D PREFIX 和 COVERAGE 策略。
- MetaSearcher 以贪心方式最多选择 k 个 peer：PREFIX 优先延长连续前缀，COVERAGE 优先增加未覆盖 key 数量。
- 同一个 V6D selector 内，每个 key 只归属第一个选中的可用 peer，避免返回重复 V6D 地址。
- max_peer_count 为 1 时保持原有单 peer 选择行为。

验证

- 相关 ASAN 测试：4/4 通过。
- KVCM 核心全量测试：121 个通过，1 个跳过。

评审重点

- PREFIX 和 COVERAGE 的增量选择及稳定排序逻辑。
- max_peer_count 为 1 时的兼容行为，以及非法参数校验。

---

Background

GetCacheLocationsByBackend currently selects at most one peer for each V6D selector. This limits cache hit coverage when blocks are distributed across multiple peers.

Implementation

- Add max_peer_count to BackendLocationSelector and require it to be greater than 0.
- Validate the parameter in CacheManager and restrict it to the V6D PREFIX and COVERAGE strategies.
- Let MetaSearcher greedily select up to k peers: PREFIX prioritizes extending the contiguous prefix, while COVERAGE prioritizes newly covered keys.
- Within one V6D selector, assign each key to the first selected available peer to avoid duplicate V6D locations.
- Preserve the existing single-peer behavior when max_peer_count is 1.

Verification

- Relevant ASAN tests: 4/4 passed.
- Full KVCM core test suite: 121 passed and 1 skipped.

Review focus

- Incremental selection and deterministic tie-breaking for PREFIX and COVERAGE.
- Compatibility when max_peer_count is 1 and validation of invalid values.

### CODE REVIEW

File: kv_cache_manager/manager/cache_manager.cc [hunk 1]
```diff
@@ -957,6 +961,19 @@ CacheManager::GetCacheLocationsByBackend(RequestContext *request_context,
             RETURN_IF_EC_NOT_OK_WITH_TYPE_LOG(
                 WARN, EC_BADARGS, BatchLocationsView, "backend selector contains duplicate backend_type");
         }
+        const bool supports_multiple_peers =
+            selector.backend_type == DataStorageType::DATA_STORAGE_TYPE_EVENT_REPORT_L2 &&
+            (selector.strategy == LocationSelectStrategy::LSS_V6D_PREFIX ||
+             selector.strategy == LocationSelectStrategy::LSS_V6D_COVERAGE);
+        if (supports_multiple_peers) {
+            selector.max_peer_count = EnvUtil::GetEnv("KVCM_V6D_MAX_PEER_COUNT", selector.max_peer_count);
```
Comment (wangxiyu191) [inline kv_cache_manager/manager/cache_manager.cc:969]: 这里在请求值已经归一化之后再次用环境变量覆盖 `max_peer_count`，但覆盖后没有重新做正值校验。

当 `KVCM_V6D_MAX_PEER_COUNT=0` 时，`SelectV6DPeers` 的循环一次也不会执行，所有 V6D 命中都会变成空结果；负数随后转换成 `size_t`，则近似取消 peer 上限。我用本 PR 新增的测试加 `--test_env=KVCM_V6D_MAX_PEER_COUNT=0` 做了复现，两个预期命中位置都实际返回了 0 locations。

建议在最终有效值确定后再统一校验，并补充 0 和负数的测试。

---

The request value is normalized before this line, but the environment variable overwrites `max_peer_count` afterward without another positive-value check.

With `KVCM_V6D_MAX_PEER_COUNT=0`, the loop in `SelectV6DPeers` never runs, so every V6D hit becomes an empty result. A negative value is later converted to `size_t`, which effectively removes the peer limit. I reproduced the zero case by running the test added in this PR with `--test_env=KVCM_V6D_MAX_PEER_COUNT=0`; both positions that were expected to hit returned zero locations.

Please validate the final effective value after applying the override, and add test coverage for zero and negative values.

File: kv_cache_manager/manager/test/cache_manager_test.cc [hunk 2]
```diff
@@ -8241,6 +8283,136 @@ TEST_F(CacheManagerTest, TestGetCacheLocationsByBackend) {
...
+    };
+
+    for (auto strategy : {LocationSelectStrategy::LSS_V6D_PREFIX, LocationSelectStrategy::LSS_V6D_COVERAGE}) {
+        check_hits(strategy, 4, 2);
+        {
+            ScopedEnv env("KVCM_V6D_MAX_PEER_COUNT", "1");
+            check_hits(strategy, 4, 1);
+        }
+        {
+            ScopedEnv env("KVCM_V6D_MAX_PEER_COUNT", "2");
```
Comment (wangxiyu191) [inline kv_cache_manager/manager/test/cache_manager_test.cc:8320]: 这个测试固定了 complete override 的语义：request 传 1、env 配 2，最终实际使用 2。也就是说，env 不仅能限制请求值，也能提高调用方明确传入的 maximum；旧 client 没有传该字段时，也会被全局切换到 multi-peer。

建议把它作为 `InstanceGroup.cache_config.default_v6d_max_peer_count`，通过现有 `UpdateInstanceGroup` 路径运行时修改。显式 request 优先；request 未指定时使用 group default；group 未配置时回落到 1。这样也比节点本地 env 更容易保持 HA 副本一致。

如果真正需要的是 server-side cap，建议使用单独的 cap 语义，而不要和 default 混在一起。

---

This test establishes complete-override semantics: request=1 and env=2 results in an effective value of 2. Therefore, the environment variable can increase the caller's explicit maximum, not only reduce it; old clients that omit the field are also globally switched to the multi-peer path.

Consider placing this setting in `InstanceGroup.cache_config.default_v6d_max_peer_count` and updating it at runtime through the existing `UpdateInstanceGroup` path. An explicit request value should win; otherwise use the group default, then fall back to 1. This would also make HA consistency easier to maintain than a node-local environment variable.

If the actual requirement is a server-side cap, it would be clearer to model that as a separate cap rather than mixing cap and default semantics.

File: kv_cache_manager/manager/meta_searcher.cc [hunk 3]
```diff
@@ -451,6 +451,112 @@ SelectV6DByIncrementalCoverage(const std::vector<bool> &base_hits,
...
+        selections.push_back(std::move(selection));
+        return selections;
+    }
+
+    std::vector<bool> combined_hits = base_hits;
+    auto remaining_candidates = remote_peer_candidates;
+    std::vector<std::string> owner_by_key(base_hits.size());
+    std::vector<V6DPeerSelection> selections;
+
+    while (selections.size() < max_peer_count) {
```
Comment (wangxiyu191) [inline kv_cache_manager/manager/meta_searcher.cc:486]: K 是否值得保留需要再权衡。

“至多选择 K 个 peer 使覆盖最多”本身是 NP-hard 的带基数约束覆盖问题；PREFIX 也有同类组合复杂度。当前 PR 采用贪心近似，但这个实现已经表现出明显的性能风险；即使按下一条建议改成更高效的索引和增量算法，只要继续保留贪心选择，评分、累计状态、停止条件和 tie-break 等复杂度仍然存在。

放开 K 后，不再需要在候选 peer 子集中求最优组合，选择逻辑会简单很多，命中覆盖也不会再被 K 截断。对应的代价是数据可能分散到更多 peer，需要测试这种情况下的真实传输性能，并在传输逻辑中处理 RDMA incast。

---

The value of retaining K needs further consideration.

Selecting at most K peers for maximum coverage is itself an NP-hard cardinality-constrained coverage problem, and PREFIX has the same kind of combinatorial complexity. The current PR uses a greedy approximation, but this implementation already shows clear performance risks. Even with a more efficient indexed and incremental implementation, retaining greedy selection still requires non-trivial scoring, accumulated state, stopping rules, and tie-breaking logic.

Without the K limit, the selector no longer needs to optimize over peer subsets, the logic becomes much simpler, and achievable cache coverage is no longer truncated. The trade-off is that data may be spread across more peers, so real transfer performance needs to be tested and RDMA incast handled in the transfer path.

File: kv_cache_manager/manager/meta_searcher.cc [hunk 4]
```diff
@@ -451,6 +451,112 @@ SelectV6DByIncrementalCoverage(const std::vector<bool> &base_hits,
...
+        if (selection.peer_addr.empty()) {
+            return {};
+        }
+        std::vector<V6DPeerSelection> selections;
+        selections.push_back(std::move(selection));
+        return selections;
+    }
+
+    std::vector<bool> combined_hits = base_hits;
+    auto remaining_candidates = remote_peer_candidates;
```
Comment (wangxiyu191) [inline kv_cache_manager/manager/meta_searcher.cc:482]: 当前路径会复制带 endpoint string 的候选表，并在每一轮重新转置、扫描候选和构造 `combined_hit`，在候选数量较多时会产生明显的 CPU、allocation 和内存开销。

本地 16K keys、256 peers、k=20 的测试中，每个 key 平均 4 个候选时 PREFIX/COVERAGE 约为 33/116 ms；平均 16 个候选时约为 192–207/370–374 ms。R 和并发较高时，CPU、allocation 和峰值内存压力可能比较明显。

可以在候选枚举时先把 endpoint 映射为紧凑的 `PeerId`，并在请求内一次构造两个只保存真实 key-peer 关系的稀疏视图：

- `candidates_by_key[key]` 保存该 key 的 `PeerId + location`，用于查找能覆盖某个 key 的 peer 以及最终 location 回填；
- `keys_by_peer[peer]` 保存该 peer 能覆盖的有序 key index，后续评分不再重新转置候选表。

选择过程中只额外维护 `selected_peer` 和当前累计命中状态。PREFIX 先找到当前 first gap，只评估 `candidates_by_key[first_gap]` 中尚未选择的 peer；对每个候选，把它的有序 `keys_by_peer` 与当前命中状态合并，计算前缀能推进到哪里，不需要为每个 peer 构造一份 `combined_hit`。COVERAGE 则一次初始化每个 peer 对未命中 key 的 gain；某个 key 首次被覆盖时，通过 `candidates_by_key[key]` 递减相关 peer 的 gain，所有增量更新合计 O(E)，每轮只需从 P 个 peer 中选择 winner。

这样可以去掉每轮的 `peer_to_indices` 转置、endpoint string map/copy、`remaining_candidates` erase 和重复位图构造，空间保持 O(N+P+E)，不会退化成 N×P 矩阵。我的本地原型通过了 40,000 次新旧差分运行；在目标场景中，完整候选阶段约快 6.6–28.5×，峰值内存约降低 8×。

这些数字只覆盖 manager 内部候选阶段，不代表整个 RPC 的加速比。

---

The current path copies the candidate table including endpoint strings, then retransposes and rescans candidates and rebuilds `combined_hit` on every round. With many candidates, this creates substantial CPU, allocation, and memory overhead.

In a local benchmark with 16K keys, 256 peers, and k=20, PREFIX/COVERAGE took about 33/116 ms with an average of 4 candidates per key, and about 192–207/370–374 ms with 16 candidates per key. At higher replication and concurrency, the CPU, allocation, and peak-memory costs may become substantial.

Intern endpoints into compact `PeerId`s during candidate enumeration and build two request-scoped sparse views containing only real key-peer edges:

- `candidates_by_key[key]` stores `PeerId + location` for finding the peers that cover a key and for final location materialization;
- `keys_by_peer[peer]` stores the ordered key indices covered by that peer, so scoring no longer retransposes the candidate table.

Selection then needs only `selected_peer` state and the accumulated hit state. PREFIX finds the current first gap and evaluates only unselected peers in `candidates_by_key[first_gap]`; for each candidate, it merges the ordered `keys_by_peer` list with the accumulated hits to determine how far the prefix advances, without allocating a per-peer `combined_hit`. COVERAGE initializes each peer's gain over currently uncovered keys once; when a key becomes covered for the first time, `candidates_by_key[key]` is used to decrement the gains of the affected peers. The total gain updates are O(E), followed by a scan of P peers per selection round.

This removes the per-round `peer_to_indices` transpose, endpoint-string maps and copies, `remaining_candidates` erases, and repeated bitmap construction. Memory remains O(N+P+E), rather than an N×P matrix. My local prototype passed 40,000 differential runs against the current behavior; in the target scenarios, the complete candidate stage was about 6.6–28.5× faster with roughly 8× lower peak memory.

These measurements cover only the manager's candidate-processing stage, not the entire RPC.

**Review by wangxiyu191 — COMMENTED** [review]
1. `KVCM_V6D_MAX_PEER_COUNT` 覆盖请求参数后没有重新校验。`env=0` 会让 V6D 结果全空，负数转成 `size_t` 后会近似取消 peer 数量限制。
2. 环境变量会无条件覆盖调用方显式传入的 `max_peer_count`。这个配置更适合作为 instance-group 级的运行时默认值：请求明确指定时使用请求值，否则使用 group default。
3. K 把 peer 选择变成了带基数约束的覆盖问题。当前贪心实现有明显的性能风险；即使优化数据结构，仍然需要维护评分、累计状态、停止条件和 tie-break。建议重新评估是否需要限制 K。放开 K 可以简化选择逻辑并避免损失命中覆盖，但需要测试多 peer 下的真实传输性能，并在传输逻辑中处理 RDMA incast。

---

1. `KVCM_V6D_MAX_PEER_COUNT` is not validated after overriding the request value. `env=0` makes all V6D results empty, while a negative value is converted to `size_t` and effectively removes the peer-count limit.
2. The environment variable unconditionally overrides an explicit `max_peer_count` from the caller. This setting would work better as an instance-group-scoped runtime default: use the request value when explicitly provided, otherwise use the group default.
3. K turns peer selection into a cardinality-constrained coverage problem. The current greedy implementation has clear performance risks, and even with better data structures it still needs scoring, accumulated state, stopping conditions, and tie-breaking. Please reconsider whether K needs to be limited. Removing the limit would simplify selection and avoid truncating cache coverage, while requiring real transfer-performance testing with more peers and RDMA-incast handling in the transfer path.
